Port output and tests to webpack

pamelafox opened the pull request:
### High-level description of change

This is a subset of PR#669 (https://github.com/Khan/live-editor/pull/669), the port to webpack/React, This PR will go into pamela-bigbranch, not into master. Subsequent PRs go also be merged into that branch, and we will not merge the full branch until everything is fully functional and tested.

This particular PR ports the output side of LiveEditor from bower to webpack. At a high level, that involves:
* Using import and export (instead of globals on window)
* Moving the copy of build assets from bower to webpack (sounds, images)
* Rewriting the demos to refer to the new bundled build files
* Rewriting the tests to use import/exports

Note that the Output JS files are all written in a large number of ways, so in some files, I exported at bottom, in others, I exported at the top - I typically picked whatever was the minimal change. I am happy to use a more consistent style, it'll probably just mean a larger diff.

### Are there performance implications for this change?

The bundle sizes are different, as there are less bundles. So that could be good or bad, depending how you measure it. 

### Have you added tests to cover this new/updated code?

No, I ported the existing output tests over. All of the output tests pass!

### Risks involved

I wrote this PR in a way that works with the webapp kake setup. It's possible that it won't be as compatible with the webpack webapp port - hard for me to say. This does get us much closer.

### Are there any dependencies or blockers for merging this?

This is not a fully functional branch. The output works, but the input needs to be ported, as there are files shared between the two.

### How can we verify that this change works?

The output tests all pass, and the demos work.

kevinbarabash commented on js/workers/pjs/jshint-worker.js:
```diff
@@ -6,15 +10,6 @@ self.onmessage = function(event) {
-            "?cachebust=" + (new Date()).toDateString());
-
-        importScripts(event.data.externalsDir +
-            "underscore/underscore.js?cachebust=" + (new Date()).toDateString());
```
Why can we get rid of these `importScripts`?

pamelafox replied:
We now load in the built worker file, so that entry point includes es5-shim and JSHint.
JSHint never used underscore. It looks like your debugger PR added underscore, and then we didn't later remove it:
https://github.com/Khan/live-editor/commit/10f0c17a3c410b9b995b9253097bf8ab8e6821a6

kevinbarabash commented on js/workers/pjs/test-worker.js:
```diff
@@ -1,39 +1,21 @@
-
-        // Bring in the testing logic
-        importScripts("../shared/output-tester.js?cachebust=" + date);
-        importScripts("./pjs-tester.js?cachebust=" + date);
```
Why are we able to get rid of all of these `importScripts`?

kevinbarabash commented on package.json:
```diff
@@ -21,11 +21,17 @@
   "devDependencies": {
     "async": "^0.9.0",
     "babel-core": "5.6.18",
```
This is a pretty old version of `babel-core`.  We should probably use the most recent version of the `6.x` line.

pamelafox replied:
Changing to "babel-core": "^6", is that right? I notice wonder-blocks using 7

kevinbarabash replied:
webapp uses 6 right now.  I think we'll want to migrate everything to 7 eventually but there were some growing pains moving from 6 to 7 in wonder-blocks that we probably want to avoid at the moment.

kevinbarabash commented on package.json:
```diff
@@ -52,7 +58,25 @@
+    "escodegen": "^1.10.0",
+    "escope": "^1.0.0",
+    "esprima": "^4.0.0",
+    "handlebars-loader": "^1.7.0",
```
Do we still need the "handlebars" dependency if we're using "handlebars-loader"?

pamelafox replied:
Good Q. We import handlebars/runtime in SQL in order to register a helper dynamically, so we still use it. FOR NOW... BUM BUM BUM.

kevinbarabash commented on tests/output/pjs/ast_transform_test.js:
```diff
@@ -1,3 +1,17 @@
+import PJSCodeInjector from "../../../js/output/pjs/pjs-code-injector.js";
+
+const getCodeFromOptions = function(code) {
```
Where is this used?  Did you maybe forget an `export`?

pamelafox replied:
Great catch. It's used a bunch later in the file. I'm changing it to import from pjs-utils instead. I'm confused as to why I pasted a copy in the file instead.

kevinbarabash commented on tests/output/pjs/test_utils.js:
```diff
@@ -43,20 +43,16 @@ var runTest = function(options) {
         });
 
-        // Switch to the Scratchpad's version
-        output.config.switchVersion(options.version);
```
Why can we get rid of this?

pamelafox replied:
Good question. I could not find any PJS tests that specified the version option, and it's not documented, and ScratchpadConfig() isn't my favorite class to have a dependency on (it's a Backbone model), so I happily did away with it here.

kevinbarabash commented on tests/output/sql/index.html:
```diff
@@ -9,22 +9,25 @@
+        <!-- Mocha, sinon, and test JS -->
         <script src="../../../node_modules/mocha/mocha.js"></script>
         <script src="../../../node_modules/expect.js/index.js"></script>
+        <script src="../../../node_modules/sinon/pkg/sinon.js"></script>
```
Were we not using sinon before?

pamelafox replied:
Good catch, we still don't use Sinon for SQL tests. This was copy/pasta, removing it!

kevinbarabash commented on tests/output/sql/test_utils.js:
```diff
@@ -86,14 +94,6 @@ var assertTest = function(options) {
-                } else {
-                    var $html = $("<div>" + errors[0].text + "</div>");
-                    expect($html.text()).to.be.equal(options.reason);
-                }
```
I'm glad this can go away now, but I don't understand why it was needed in this file in the first place since it's used by the sql tests.

pamelafox replied:
Generally the test_utils files have a lot in common with each other across pjs/webpage/sql, so I imagine the original was copy/pasta and we didn't notice that there was unused code. (If we had a coverage tool, we might identify such areas)

kevinbarabash commented on tests/output/pjs/iframe.html:
```diff
@@ -10,24 +10,23 @@
-                "async_test.js",
-                "jshint_test.js",
-                "output_test.js",
+                "../../../build/js/live-editor.tests_output_pjs_jshint.js",
```
I'm guessing this change is because we can't run the source code inside phantomjs anymore, right?

pamelafox replied:
It was easier for each of the tests to be a built file, including all its dependencies. Otherwise I'd need to do a script include that exposed the relevant classes, right?
(Perhaps there's another way that I'm missing)

kevinbarabash replied:
The changes are fine.  I'm just trying to understand the rationale behind the changes. 🙂

kevinbarabash commented on tests/output/webpage/test_utils.js:
```diff
@@ -1,29 +1,34 @@
 
-    var displayTitle = options.title +
-        " (Version: " + options.version + ")";
+export function createLiveEditorOutput(extraOptions) {
```
FYI, you can use default args now, e.g.
```
function createLiveEditorOutput(extraOptions = {})
```

pamelafox replied:
Yay! JS and Python had a beautiful ES6 baby!

kevinbarabash replied:
I think we're okay in this case, but we'd definitely want to avoid this in python.

kevinbarabash commented on webpack.config.js:
```diff
@@ -0,0 +1,104 @@
+                use: [
+                    {
+                        loader: "expose-loader",
+                        options: "_",
```
Can you add a comment about why we're doing this for `underscore`?

kevinbarabash commented on webpack.config.js:
```diff
@@ -0,0 +1,104 @@
+    module: {
+        rules: [
+            {
+                test: /\.(js|jsx)$/,
```
`/\.js$/` should be okay since I think we want to use `.js` for files containing JSX as well.

kevinbarabash commented on webpack.config.js:
```diff
@@ -0,0 +1,104 @@
+        rules: [
+            {
+                test: /\.(js|jsx)$/,
+                exclude: /(node_modules|external|tests|\.(test)\.(js|jsx)$)/,
```
Does `/tests/*.test.js` need to be excluded?  It looks like our test files actually include `_test.js` and need to be transformed anyways.

pamelafox replied:
Good point, removing tests from exclude.

pamelafox replied:
Ah: the tests in ast_transform_test fail when babel transforms the code, due to the whitespace issues it introduces. I've excluded it for now.

kevinbarabash commented on webpack.config.js:
```diff
@@ -0,0 +1,104 @@
+const CopyWebpackPlugin = require("copy-webpack-plugin");
+
+const config = {
+    performance: {hints: false},
```
I'm not familiar with this option, what does it do?

kevinbarabash commented on webpack.config.js:
```diff
@@ -0,0 +1,104 @@
+        output_pjs_deps: [
+            "./external/processing-js/processing.js",
+            "./external/jshint/jshint.js",
+        ],
```
This is not very webpack-y but it does what we need.

MiguelCastillo replied:
interesting - @kevinbarabash which part is not this not very `webpack-y`?

pamelafox replied:
I think the fact that I'm making a separate bundle for the dependencies, and the dependencies bundle exposes its stuff as globals. The output file makes that more clear, as you can see I include two bundles:
https://github.com/Khan/live-editor/pull/686/files#diff-315326e54ee98fc7fd456045df327d7d

kevinbarabash commented on webpack.config.js:
```diff
@@ -0,0 +1,104 @@
+    ],
+    output: {
+        path: path.resolve(__dirname, "build/js"),
+        filename: "live-editor.[name].js",
```
nice

kevinbarabash commented:
I'm pretty sure that we have automated tests for the infinite loop checker, but if my memory's failed me and we don't it'd be good to manually check that that still works.

pamelafox commented on demos/simple/index.html:
```diff
@@ -64,7 +64,7 @@ <h1>Live Editor Example</h1>
         editorHeight: "80%",
         autoFocus: true,
-        workersDir: "../../build/workers/",
+        workersDir: "../../build/",
```
The workers are now generated as webpack entry point, so they end up in builld/js like all the others. I could get them back in workers by changing the output option to modify the folder of the built bundles, but I haven't seen a need for workers being in their own folder yet.

pamelafox commented on js/output/pjs/pjs-code-injector.js:
```diff
@@ -1076,13 +1091,16 @@ class PJSCodeInjector {
                         var audio = document.createElement("audio");
                         var parts = filename.split("/");
 
-                        var group = _.findWhere(OutputSounds[0].groups, { groupName: parts[0] });
```
Yes, I think so, given we only use parts of it in webapp, and that many things are now available in ES6. I did originally make a bunch of lodash/underscore removal related changes in my big branch, but then realized that wasn't the P1 change we needed and reverted them. I kept the change here since this exported code should have minimal dependencies.

pamelafox commented on js/output/pjs/pjs-output.js:
```diff
@@ -1,4 +1,14 @@
-window.PJSOutput = Backbone.View.extend({
+const $ = require("jquery");
```
Changing to imports - that's because I was using require() at the pre-React stage of the PR and I copy-pasta'd.

pamelafox commented on js/output/pjs/pjs-output.js:
```diff
@@ -17,7 +27,7 @@ window.PJSOutput = Backbone.View.extend({
 
         this.tester = new PJSTester(_.extend(options, {
-            workerFile: "pjs/test-worker.js"
+            workerFile: "js/live-editor.test_worker.js",
```
The worker files are now built files (entry points) which contain all of their requirements. The new filename is the built file.

pamelafox replied:
Actually, I'm changing it to use a separate webapp config for the workers (as the externals are different), so I'll change it back to /workers.

pamelafox commented on js/output/pjs/pjs-utils.js:
```diff
@@ -1,4 +1,4 @@
-window.PJSUtils = {
+const PJSUtils = {
```
I've identified three files in this diff that would be better done that way:
- Pjs-utils
- Pjs-ast-transforms
- Ast-builder

I'm happy to make that change in this diff, if you're OK with the larger diff. (Sometimes big whitespace changes throw off the diffing so the PR looks bigger, I don't know how it'll render that change). Those files are minimally changed otherwise.

kevinbarabash replied:
Sorry for using imprecise language... I meant "future PR" when I said "second pass".

pamelafox replied:
Ok, noting as a task for future PR.

pamelafox commented on js/output/shared/output-tester.js:
```diff
@@ -136,7 +142,7 @@ OutputTester.prototype = {
             return true;
         }
-
+        code = code.replace(/\$\._/g, "i18n._");
```
Good call, adding the comment.

pamelafox commented on js/output/pjs/pjs-code-injector.js:
```diff
@@ -1045,6 +1058,8 @@ class PJSCodeInjector {
         // TODO(kevinb) generate this code once (once webpack is in place)
         helperCode += `var resources = ${JSON.stringify(resources)};\n`;
+        // We rename it to OutputSounds2 so that webpack doesn't rewrite it
+        helperCode += `var OutputSounds2 = ${JSON.stringify(OutputSounds)};\n`;
```
Good question. It doesn't overwrite this name, it rewrites the later reference, on line 1103 (the object we call findWhere on). I'll update the comment.

pamelafox replied:
You likely have a better idea for how to do this, given your TODO about improving once we use webapp. Given this is currently not in use by webapp, my tendency is to defer that.
(But let me know if you have a better quick fix)

pamelafox commented on tests/output/webpage/index.html:
```diff
@@ -9,24 +9,24 @@
+        <!-- Mocha, sinon, and test JS -->
         <script src="../../../node_modules/mocha/mocha.js"></script>
         <script src="../../../node_modules/expect.js/index.js"></script>
+        <script src="../../../node_modules/sinon/pkg/sinon.js"></script>
```
Removing this copy/pasta sinon.js as well

pamelafox commented:
Thank you for the thorough review! So much helpful feedback.

I've addressed the comments and made one significant change as to how the code brings in _ and i18n. The issue before was that test-worker.js brought in pjs-tester.js which relied on those 2, and I had to do various things to make sure test-worker would always bring in _ and i18n locally and webapp.

I decided a cleaner approach is to have a separate webapp config file for the workers. That makes sense, I think, because the externals are different for local build vs worker build vs webapp build.

Specifically:
* For local demos and tests, the bundles currently import _ from lodash and actually include it in the bundles. They also import i18n from the local version and actually include it in the bundles.
* For workers, the bundles need to always have _ and i18n included, as they don't have access to globals. It's okay that test-worker bundle always brings in the local i18n versus the webapp i18n, as it doesn't really utilize it. (We previously mocked it inside test-worker).
* For webapp (in future PR), the bundles treat _ and i18n as externals, and they do *not* include them, since webapp exposes its own versions of them.

Hm. Okay, I must admit, I now realize that the worker bundles could be in the original config, as they *currently* share the same configuration options and externals. They must have been different at some point in my refactoring process. I think the workers still make sense in a separate config, as then it's easy to change some options about them, but it seems like I could move them back if desired.

Time for a brain break :)

pamelafox commented:
Okay, I'll keep workers separate as it does convey the difference in environment, and it might make future differences easier. Merging into my branch (not master!). 

My next PR will likely be the webpackification of the rest of the files (which is definitely a word).

Oh, and re earlier question: loop checking still seems to work fine, in browser and tests.
